Refactor UXTO Validation in Output Manager into async protocol

philipr-za opened the pull request:
## Description
UTXO validation in the Output Manager currently has a few elements that are synchronous relative to the main Select! Loop in the service. This meant that when there were a large number of outputs in the wallet that this could bog down the entire service.

This PR refactors the UTXO validation process into a fully asynchronous and more modular Protocol pattern that runs in its own task. The API to the process is also updated to allow clients to select the appropriate Retry strategy. The automatic Utxo Validation of Unspent outputs was removed and must now be done explicitly but the client so they have more control over when it happens. The Automatic Invalid UTXO validation is still in place but instead of repeating until success it is reduced to 5 attempts as it is not critical.

It was also found that there was one synchronous Mutex being used in the service outside of a `spawn_blocking(…)` context which could cause problems with the Tokio runtime so that has been updated to use an asynchronous Mutex.

The Output Manager was also updated to use to the Tokio Broadcast channel and the tari_broadcast_channel dependency has been removed from the Wallet crate.

A test has been added to fully test the wallet_ffi callback handler which ingests the new events the OMS is emitting.

A bunch of Clippy fixes

Closes https://github.com/tari-project/tari/issues/1968
Closes https://github.com/tari-project/tari/issues/1969

## How Has This Been Tested?
Tests updated and a new test of the Callback Handler provided

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

sdbondi commented on base_layer/wallet/src/output_manager_service/protocols/utxo_validation_protocol.rs:
```diff
@@ -0,0 +1,497 @@
+                .event_publisher
+                .send(OutputManagerEvent::UtxoValidationTimedOut(self.id))
+                .map_err(|e| {
+                    trace!(
+                        target: LOG_TARGET,
+                        "Error sending event, usually because there are no subscribers: {:?}",
+                        e
+                    );
```
nit: The **only** way that the broadcast send can fail is if there are no subscribers. Also, `e` just wraps what was sent `SendError(pub T)` and (perhaps incorrectly) does not implement `Error`.

```rust
/// A **send** operation can only fail if there are no active receivers,
/// implying that the message could never be received. The error contains the
/// message being sent as a payload so it can be recovered.
```

Something like this may be more appropriate: 
```suggestion
                    trace!(
                        target: LOG_TARGET,
                        "Did not send event `{:?}` because there are no subscribers",
                        e.0
                    );
```

philipr-za replied:
True

sdbondi commented on base_layer/wallet/src/output_manager_service/service.rs:
```diff
@@ -173,15 +165,16 @@ where
         pin_mut!(base_node_response_stream);
 
-        let mut utxo_query_timeout_futures: FuturesUnordered<BoxFuture<'static, u64>> = FuturesUnordered::new();
+        let mut utxo_validation_handles: FuturesUnordered<JoinHandle<Result<u64, OutputManagerProtocolError>>> =
```
nit: If you don't feel like maintaining the type signature for this in two places, you can just remove it here. The `handle_request` function will tell rust what type it is.

philipr-za replied:
True and I hear you on slightly less maintenance but personally I feel this is more readable.

sdbondi reviewed (dismissed):
LGTM

philipr-za committed:
Refactor UXTO Validation in Output Manager into async protocol

UTXO validation in the Output Manager currently has a few elements that are synchronous relative to the main Select! Loop in the service. This meant that when there were a large number of outputs in the wallet that this could bog down the entire service.

This PR refactors the UTXO validation process into a fully asynchronous and more modular Protocol pattern that runs in its own task. The API to the process is also updated to allow clients to select the appropriate Retry strategy. The automatic Utxo Validation of Unspent outputs was removed and must now be done explicitly but the client so they have more control over when it happens. The Automatic Invalid UTXO validation is still in place but instead of repeating until success it is reduced to 5 attempts as it is not critical.

It was also found that there was one synchronous Mutex being used in the service outside of a `spawn_blocking(…)` context which could cause problems with the Tokio runtime so that has been updated to use an asynchronous Mutex.

The Output Manager was also updated to use to the Tokio Broadcast channel and the tari_broadcast_channel dependency has been removed from the Wallet crate.

A test has been added to fully test the wallet_ffi callback handler which ingests the new events the OMS is emitting.

A bunch of Clippy fixes
